srml/staking/src/lib.rs

```diff
@@ -661,58 +892,68 @@ impl<T: Trait> Module<T> {
 	///
-	/// @returns the new SlotStake value.
+	/// Returns the new SlotStake value.
 	fn select_validators() -> BalanceOf<T> {
 		// Map of (would-be) validator account to amount of stake backing it.
-		
+
 		let rounds = || <ValidatorCount<T>>::get() as usize;
 		let validators = || <Validators<T>>::enumerate();
 		let nominators = || <Nominators<T>>::enumerate();
-		let stash_of = |w| Self::stash_balance(&w);
+		let stash_of = |w: &T::AccountId| -> BalanceOf<T> { Self::stash_balance(w) };
 		let min_validator_count = Self::minimum_validator_count() as usize;
-		let elected_candidates = phragmen::elect::<T, _, _, _, _>(
+		let maybe_elected_candidates = elect::<T, _, _, _, _>(
 			rounds,
 			validators,
 			nominators,
 			stash_of,
-			min_validator_count
+			min_validator_count,
+			ElectionConfig::<BalanceOf<T>> {
+				equalise: false,
+				tolerance: <BalanceOf<T>>::sa(10 as u64),
+				iterations: 10,
+			}
 		);
 
-		// Figure out the minimum stake behind a slot.
-		let slot_stake = elected_candidates
-			.iter()
-			.min_by_key(|c| c.exposure.total)
-			.map(|c| c.exposure.total)
-			.unwrap_or_default();
-		<SlotStake<T>>::put(&slot_stake);
-
-		// Clear Stakers and reduce their slash_count.
-		for v in <session::Module<T>>::validators().iter() {
-			<Stakers<T>>::remove(v);
-			let slash_count = <SlashCount<T>>::take(v);
-			if slash_count > 1 {
-				<SlashCount<T>>::insert(v, slash_count - 1);
+		if let Some(elected_candidates) = maybe_elected_candidates {
+			// Clear Stakers and reduce their slash_count.
+			for v in <session::Module<T>>::validators().iter() {
+				<Stakers<T>>::remove(v);
+				let slash_count = <SlashCount<T>>::take(v);
+				if slash_count > 1 {
+					<SlashCount<T>>::insert(v, slash_count - 1);
+				}
 			}
-		}
 
-		// Populate Stakers.
-		for candidate in &elected_candidates {
-			<Stakers<T>>::insert(candidate.who.clone(), candidate.exposure.clone());
-		}
+			// Populate Stakers and figure out the minimum stake behind a slot.
+			let mut slot_stake = elected_candidates[0].exposure.total;
+			for c in &elected_candidates {
+				if c.exposure.total < slot_stake {
+					slot_stake = c.exposure.total;
+				}
+				<Stakers<T>>::insert(c.who.clone(), c.exposure.clone());
+			}
+			<SlotStake<T>>::put(&slot_stake);
 
-		// Set the new validator set.
-		<session::Module<T>>::set_validators(
-			&elected_candidates.into_iter().map(|i| i.who).collect::<Vec<_>>()
-		);
-		
-		slot_stake
+			// Set the new validator set.
+			<session::Module<T>>::set_validators(
+				&elected_candidates.into_iter().map(|i| i.who).collect::<Vec<_>>()
+			);
+
+			slot_stake
+		} else {
+			// There were not enough candidates for even our minimal level of functionality.
+			// This is bad.
+			// We should probably disable all functionality except for block production
+			// and let the chain keep producing blocks until we can decide on a sufficiently
+			// substantial set.
+			Self::slot_stake()
```

Comment:
One suggestion from @rphmeier which I think went missing here: Even when we are naively choosing the previous set, `take()` only `validator_count` from it. It might be that we had more than what we need and at least we can kick out the extra one.

Reply:
that would only happen if `minimum_validator_count` increased. which seems like even more of an edge case than this. it also substantially complicates things.